src/new-storage/Storage.js

```diff
@@ -173,55 +152,191 @@ class Storage extends EventEmitter {
-        const stream2 = this._queryWithStreamingResults(query2, queryParams2)
-        const stream3 = this._queryWithStreamingResults(query3, queryParams3)
-
-        pump(
-            merge2(stream1, stream2, stream3),
-            resultStream,
-            (err) => {
-                if (err) {
-                    console.error('pump finished with error', err)
-                    resultStream.push(null)
+        const query = 'SELECT payload FROM stream_data_new WHERE '
+            + 'stream_id = ? AND partition = ? AND bucket_id IN ? AND ts >= ? AND ts <= ?'
+
+        Promise.all([
+            this.bucketManager.getLastBuckets(streamId, partition, 1, fromTimestamp),
+            this.bucketManager.getLastBuckets(streamId, partition, 1, toTimestamp),
+        ]).then((results) => {
+            const date1 = results[0][0].dateCreate
+            const date2 = results[1][0].dateCreate
+            return [Math.min(date1, date2), Math.max(date1, date2)]
+        }).then(([startBucketDate, endBucketDate]) => {
+            return this.bucketManager.getBucketsByTimestamp(streamId, partition, startBucketDate, endBucketDate)
+        }).then((buckets) => {
+            const bucketsForQuery = []
+
+            for (let i = 0; i < buckets.length; i++) {
+                const bucket = buckets[i]
+                bucketsForQuery.push(bucket.id)
+            }
+
+            const queryParams = [streamId, partition, bucketsForQuery, fromTimestamp, toTimestamp]
+            const cassandraStream = this._queryWithStreamingResults(query, queryParams)
+
+            return pump(
+                cassandraStream,
+                resultStream,
+                (err) => {
+                    if (err) {
+                        console.error('pump finished with error', err)
+                        resultStream.push(null)
+                    }
                 }
+            )
+        })
+            .catch((e) => {
+                console.warn(e)
+                resultStream.push(null)
+            })
+
+        return resultStream
+    }
+
+    _fetchBetweenMessageRefsForPublisher(streamId, partition, fromTimestamp, fromSequenceNo, toTimestamp, toSequenceNo, publisherId, msgChainId) {
+        const resultStream = this._createResultStream()
+
+        const query1 = 'SELECT payload FROM stream_data_new WHERE stream_id = ? AND partition = ? AND bucket_id IN ? AND ts = ? AND sequence_no >= ? AND publisher_id = ? '
+            + 'AND msg_chain_id = ? ALLOW FILTERING'
+        const query2 = 'SELECT payload FROM stream_data_new WHERE stream_id = ? AND partition = ? AND bucket_id IN ? AND ts > ? AND ts < ? AND publisher_id = ? '
+            + 'AND msg_chain_id = ? ALLOW FILTERING'
+        const query3 = 'SELECT payload FROM stream_data_new WHERE stream_id = ? AND partition = ? AND bucket_id IN ? AND ts = ? AND sequence_no <= ? AND publisher_id = ? '
+            + 'AND msg_chain_id = ? ALLOW FILTERING'
+
+        // TODO replace with allSettled
```

Comment:
About TODO: I mean if we can't effectively fetch the buckets, can we really carry on meaningfully with the operation?

Reply:
My point being that I think `Promise.all` is pretty reasonable here. If we can't fulfil the query in its entirety maybe it is reasonable to just error as we do now.